### PR TITLE
[RFR] Decoupling allowEmpty prop from reference selectors

### LINKED ISSUE #953: Update docs ReferenceField about option allowEmpty
There is no mention about option `allowEmpty` in docs about ReferenceField https://marmelab.com/admin-on-rest/Fields.html#referencefield
We can see progress bar without this option. 
So, if this fix already done or will be done in PR https://github.com/marmelab/admin-on-rest/pull/873 you can close it. Thank you

### LINKED ISSUE #1153: allowEmpty mandatory on Create form using a ReferenceInput
**What you were expecting:**
Using `ReferenceInput` inside a `Create` `Simpleform` to work without having to use the `allowEmpty` prop

**What happened instead:**
The component is disabled

**Related code:**
https://codesandbox.io/s/52o4m4pxo4

**Other information:**
To be able to enter a resource with a mandatory `ReferenceInput`, there is no other choice than to use the allowEmpty prop in the `Create `config. Which is very contradictory, specially together with `validation={{ required: true }}`

Since there is no reference id yet, and allowEmpty is false, this piece of code is executed in `ReferenceInput`

```
            if (!referenceRecord && !allowEmpty) {
                return _react2.default.createElement(_Labeled2.default, {
                    label: typeof label === 'undefined' ? 'resources.' + resource + '.fields.' + source : label,
                    source: source,
                    resource: resource
                });
            }
```
and the dropdown is deactivated

image

Could we change this condition and not do this if we are on a create form maybe ?

**Environment**

* Admin-on-rest version: 1.3.1
* React version: 15.6.1
* Browser: Chrome

### PR DESCRIPTION
**Pull request by alexisjanvier**
Fix #953 
Fix #1153 

`allowEmpty` should only be used to add or not an empty choice in the choices list, with a value of null.
It must not interfere with the fetch of references.

The reference selectors must indicate either that the reference fetch is in progress, that the reference fetch has failed, or display the *choice component* with choices based on the fetched references.

### todo
- [x] Display of a spinner during references fetch
fetchok

- [x] Display fetch error if it exists
fetchko

- [x] Render children component after fetch with allowEmpty prop binded
- [x] Update documentation about `allowEmpty`

### CODE REVIEW

File: src/mui/input/ReferenceInput.js [hunk 1]
```diff
@@ -164,20 +175,24 @@ export class ReferenceInput extends Component {
             onChange,
             children,
             meta,
+            record,
         } = this.props;
-        if (!referenceRecord && !allowEmpty) {
+        if (!matchingReferences.length) {
```
Comment (fzaninotto) [inline src/mui/input/ReferenceInput.js]: not good: for instance, for an autocomplete, if I type a radom string, there will be no matching reference, even once the fetch has succeeded

File: src/mui/input/ReferenceInput.js [hunk 2]
```diff
@@ -218,6 +233,7 @@ ReferenceInput.propTypes = {
     meta: PropTypes.object,
     onChange: PropTypes.func,
     perPage: PropTypes.number,
+    record: PropTypes.object,
```
Comment (fzaninotto) [inline src/mui/input/ReferenceInput.js]: unused

File: src/reducer/admin/references/possibleValues.js [hunk 3]
```diff
@@ -9,6 +12,11 @@ export default (previousState = initialState, { type, payload, meta }) => {
                 ...previousState,
                 [meta.relatedTo]: payload.data.map(record => record.id),
             };
+        case CRUD_GET_MATCHING_FAILURE:
+            return {
+                ...previousState,
+                [meta.relatedTo]: { error: payload.error },
```
Comment (alexisjanvier) [inline src/reducer/admin/references/possibleValues.js]: I'm not very fan of having either an object or an array for this state value... Does anyone have a better solution in mind?

File: src/mui/input/ReferenceArrayInput.js [hunk 4]
```diff
@@ -172,6 +176,43 @@ export class ReferenceArrayInput extends Component {
...
+                        ) : (
+                            label
+                        )
+                    }
+                    errorText={matchingReferences.error}
+                />
+            );
+        }
+
         if (!(referenceRecords && referenceRecords.length > 0) && !allowEmpty) {
```
Comment (alexisjanvier) [inline src/mui/input/ReferenceArrayInput.js]: IMO, `refrenceRecords` should be deprecated. @fzaninotto @djhi are you agree with that ?

Comment (djhi) [reply src/mui/input/ReferenceArrayInput.js]: Can you elaborate ?

Comment (alexisjanvier) [reply src/mui/input/ReferenceArrayInput.js]: Here, we test if we have the referenceIds data in the state resources
```
const referenceIds = props.input.value || [];
const data = state.admin.resources[props.reference].data;
referenceRecords: referenceIds.reduce((references, referenceId) => {
            if (data[referenceId]) {
                references.push(data[referenceId]);
            }
            return references;
        }, []),
```
If not, and if `allowEmpty` is not `true`, we show a disabled input. 
I think we don't need to do this anymore (because we test if value is required with `validate`, and not with `allowEmpty` prop), so we should remove this `recordReference` from props.

File: src/mui/input/ReferenceArrayInput.js [hunk 5]
```diff
@@ -182,8 +185,29 @@ export class ReferenceArrayInput extends Component {
                             label
                         )
                     }
-                    source={source}
-                    resource={resource}
+                >
+                    <span style={progessContainerStyle}>
```
Comment (djhi) [inline src/mui/input/ReferenceArrayInput.js]: Why not extracting completely this component as `ReferenceLoadingProgress` and put the styles in it ?

File: src/reducer/admin/references/possibleValues.js (hunk 3 above)
Comment (djhi) [reply src/reducer/admin/references/possibleValues.js]: Add another property ?

```js
[`${meta.relatedTo}`_error]
```

File: docs/Tutorial.md [hunk 6]
```diff
@@ -249,11 +249,12 @@ const PostTitle = ({ record }) => {
     return <span>Post {record ? `"${record.title}"` : ''}</span>;
 };
 
+const required = value => (value ? undefined : 'Required');
 export const PostEdit = (props) => (
     <Edit title={<PostTitle />} {...props}>
         <SimpleForm>
             <DisabledInput source="id" />
-            <ReferenceInput label="User" source="userId" reference="users">
+            <ReferenceInput label="User" source="userId" reference="users" validate={[required]}>
```
Comment (fzaninotto) [inline docs/Tutorial.md]: I think you must now extract the validation array, too, due to a redux-form BC break. Ping @djhi

File: src/mui/input/ReferenceArrayInput.js (hunk 5 above)
Comment (fzaninotto) [reply src/mui/input/ReferenceArrayInput.js]: 👍

File: src/mui/input/ReferenceArrayInput.js [hunk 7]
```diff
@@ -182,8 +185,29 @@ export class ReferenceArrayInput extends Component {
...
+                </Labeled>
+            );
+        }
+
+        if (matchingReferences.error) {
+            return (
+                <TextField
+                    disabled={true}
+                    hintText={
+                        typeof label === 'undefined' ? (
```
Comment (fzaninotto) [inline src/mui/input/ReferenceArrayInput.js]: missing call to translation helper

File: src/mui/input/ReferenceArrayInput.js [hunk 8]
```diff
@@ -182,8 +185,29 @@ export class ReferenceArrayInput extends Component {
...
+                            `resources.${resource}.fields.${source}`
+                        ) : (
+                            label
+                        )
+                    }
+                    errorText={matchingReferences.error}
```
Comment (fzaninotto) [inline src/mui/input/ReferenceArrayInput.js]: missing call to translation helper

File: src/mui/input/ReferenceArrayInput.js [hunk 9]
```diff
@@ -172,7 +175,7 @@ export class ReferenceArrayInput extends Component {
             );
         }
 
-        if (!(referenceRecords && referenceRecords.length > 0) && !allowEmpty) {
```
Comment (fzaninotto) [inline src/mui/input/ReferenceArrayInput.js]: You miss a case where the current record references non-existent records in the distant resource. E.g. an article with a list of author_ids, some of which don't exist in the Author resource. The previous code was there to avoid the app to break in that case (and we should keep this security).

File: src/mui/input/ReferenceInput.js [hunk 10]
```diff
@@ -157,15 +169,21 @@ export class ReferenceInput extends Component {
...
             basePath,
             onChange,
             children,
             meta,
         } = this.props;
-        if (!referenceRecord && !allowEmpty) {
+
+        if (React.Children.count(children) !== 1) {
+            throw new Error(
+                '<ReferenceInput> only accepts a single child (like <SelectInput>)'
```
Comment (fzaninotto) [inline src/mui/input/ReferenceInput.js]: s/ (like <SelectInput>)//

File: src/mui/input/ReferenceInput.js [hunk 11]
```diff
@@ -175,8 +193,29 @@ export class ReferenceInput extends Component {
...
+                    </span>
+                </Labeled>
+            );
+        }
+
+        if (matchingReferences.error) {
+            return (
+                <TextField
+                    disabled={true}
+                    hintText={
```
Comment (fzaninotto) [inline src/mui/input/ReferenceInput.js]: missing call to translation helper. Same for error.

File: src/mui/input/ReferenceInput.js [hunk 12]
```diff
@@ -157,15 +169,21 @@ export class ReferenceInput extends Component {
...
             source,
-            referenceRecord,
             allowEmpty,
             matchingReferences,
             basePath,
             onChange,
             children,
             meta,
         } = this.props;
-        if (!referenceRecord && !allowEmpty) {
```
Comment (fzaninotto) [inline src/mui/input/ReferenceInput.js]: Same remark as for the `ReferenceArrayInput`, the app musn't break if the `referenceRecord` doesn't exist.

File: src/reducer/admin/references/possibleValues.js (hunk 3 above)
Comment (fzaninotto) [reply src/reducer/admin/references/possibleValues.js]: We already use this pattern otherwise, not a big deal IMO

File: src/reducer/admin/references/possibleValues.js [hunk 13]
```diff
@@ -20,17 +28,23 @@ export const getPossibleReferences = (
...
-        referenceSource
-    ]
-        ? Array.from(state.admin.references.possibleValues[referenceSource])
-        : [];
+    if (!state.admin.references.possibleValues[referenceSource]) {
+        return null;
+    }
+
+    if (state.admin.references.possibleValues[referenceSource].error) {
+        return state.admin.references.possibleValues[referenceSource];
```
Comment (fzaninotto) [inline src/reducer/admin/references/possibleValues.js]: why skip the special map / filter processing of the possible values in that case?

File: docs/Tutorial.md (hunk 6 above)
Comment (djhi) [reply docs/Tutorial.md]: Probably but doesn't validate support a single function too ?

`validate={required}`

File: src/mui/input/ReferenceArrayInput.js (hunk 7 above)
Comment (djhi) [reply src/mui/input/ReferenceArrayInput.js]: Don't forget to the `_` option

File: src/reducer/admin/references/possibleValues.js (hunk 3 above)
Comment (alexisjanvier) [reply src/reducer/admin/references/possibleValues.js]: Okay, I'll leave it like that.

File: src/reducer/admin/references/possibleValues.js (hunk 13 above)
Comment (alexisjanvier) [reply src/reducer/admin/references/possibleValues.js]: Because this case corresponds to a fetch error return

File: src/mui/input/ReferenceArrayInput.js [hunk 14]
```diff
@@ -278,13 +267,12 @@ function mapStateToProps(state, props) {
...
+const enhance = compose(
+    connect(mapStateToProps, {
+        crudGetMany: crudGetManyAction,
+        crudGetMatching: crudGetMatchingAction,
+    }),
+    translate
+);
 
-export default ConnectedReferenceInput;
+export default enhance(ReferenceArrayInput);
```
Comment (djhi) [inline src/mui/input/ReferenceArrayInput.js]: The exported component needs to define its default props, with `addField: true` in it

File: src/mui/input/ReferenceInput.js [hunk 15]
```diff
@@ -288,13 +263,12 @@ function mapStateToProps(state, props) {
...
+const enhance = compose(
+    connect(mapStateToProps, {
+        crudGetOne: crudGetOneAction,
+        crudGetMatching: crudGetMatchingAction,
+    }),
+    translate
+);
 
-export default ConnectedReferenceInput;
+export default enhance(ReferenceInput);
```
Comment (djhi) [inline src/mui/input/ReferenceInput.js]: The exported component needs to define its default props, with addField: true in it

**Comment by alexisjanvier** [comment]
@fzaninotto @djhi review apllied

File: src/mui/input/ReferenceInput.js [hunk 16]
```diff
@@ -157,28 +160,37 @@ export class ReferenceInput extends Component {
...
-            referenceRecord,
             allowEmpty,
             matchingReferences,
             basePath,
             onChange,
             children,
             meta,
+            translate,
         } = this.props;
-        if (!referenceRecord && !allowEmpty) {
```
Comment (djhi) [inline src/mui/input/ReferenceInput.js]: Aren't we still missing that case ?

**Comment by djhi** [comment]
I think we are missing some cases here:

`ReferenceInput`: If the currently referenced resource cannot be fetched for any reason, we should display that information without preventing edition.

`ReferenceArrayInput`: If any of the currently referenced resources cannot be fetched for any reason, we should display that information for each of them (even all if needed) without preventing edition.

Ping @fzaninotto

File: src/mui/input/ReferenceInput.js (hunk 16 above)
Comment (fzaninotto) [reply src/mui/input/ReferenceInput.js]: What we should do: 

- Display an error message saying that the reference wasn't found in the API
- display the input with an empty value. It should be functional (i.e. allow to set a new value)

File: src/mui/input/ReferenceInput.js [hunk 17]
```diff
@@ -157,28 +160,37 @@ export class ReferenceInput extends Component {
...
+
+        if (React.Children.count(children) !== 1) {
+            throw new Error('<ReferenceInput> only accepts a single child');
+        }
+
+        if (matchingReferences === null) {
+            return <ReferenceLoadingProgress label={translatedLabel} />;
+        }
+
+        if (error) {
```
Comment (fzaninotto) [inline src/mui/input/ReferenceInput.js]: Same as above, please handle the case where the `matchingReferences` are loaded but not the `referenceRecord`.

File: src/mui/input/ReferenceArrayInput.js [hunk 18]
```diff
@@ -172,20 +176,32 @@ export class ReferenceArrayInput extends Component {
...
+        if (
+            !error &&
+            input.value &&
+            input.value.length > 0 &&
+            !referenceRecords.length
+        ) {
+            error = translate('aor.input.references.missing');
+        }
+
+        if (matchingReferences === null) {
```
Comment (fzaninotto) [inline src/mui/input/ReferenceArrayInput.js]: why don't you do that *before* building the error message? It would provide an early return.

File: src/mui/input/ReferenceInput.js [hunk 19]
```diff
@@ -157,28 +160,37 @@ export class ReferenceInput extends Component {
...
+                : label;
+        const translatedLabel = translate(finalLabel, { _: finalLabel });
+        const error =
+            matchingReferences && matchingReferences.error
+                ? translate(matchingReferences.error, {
+                      _: matchingReferences.error,
+                  })
+                : null;
+
+        if (React.Children.count(children) !== 1) {
```
Comment (fzaninotto) [inline src/mui/input/ReferenceInput.js]: Same as above, please move up to return earlier.

File: src/mui/input/ReferenceInput.js [hunk 20]
```diff
@@ -157,28 +160,37 @@ export class ReferenceInput extends Component {
...
+            throw new Error('<ReferenceInput> only accepts a single child');
+        }
+
+        if (matchingReferences === null) {
```
Comment (fzaninotto) [inline src/mui/input/ReferenceInput.js]: I worry about another case:

1. The input displays and loads matching references => `matchingReferences` is null
2. The API returns an error => `matchingReferences` is `{ error: 'toto' }`
3. The user clicks "Refresh", which re-triggers the 2 fetch actions. But at that time, the value of `matchingReferences` is still an error, so the loader won't show...

File: src/mui/input/ReferenceInput.js (hunk 16 above)
Comment (djhi) [reply src/mui/input/ReferenceInput.js]: The error shouldn't be displayed by the `ReferenceInput` itself right ? It should just be passed to the child as `ReferenceInput` don't actually render anything.

Comment (fzaninotto) [reply src/mui/input/ReferenceInput.js]: You're right

**Comment by fzaninotto** [comment]
also, please consider #854

**Comment by alexisjanvier** [comment]
@fzaninotto #854 is about `<Reference*Field>` not `<Reference*Input>`. This should be adressed in another PR.

**Comment by alexisjanvier** [comment]
Here a summary of the new behavior of reference inputs

## `<ReferenceInput>`

#### A spinner is displayed as long as:
 * we search for matching references and there are no reference already linked to the ressource,
 * we search for matching references and also data for a reference already linked to the ressource.
capture d ecran de 2018-02-12 16-26-30

#### An error is displayed if:
 * we have a server error return on the search of matching references and that there are no reference already linked to the ressource,
 * we have a server error return on the search of matching references and the data of reference already associated with the object are not found.
capture d ecran de 2018-02-12 16-28-21

#### A warning is displayed with a children's error when : 
 * we have a server error return on the search of matching references but we found data for reference already associated with the object
 * we didn't find any data for a reference already associated with the object, but we have a valid return (even if empty) from the server on the search of matching references.
capture d ecran de 2018-02-12 16-29-25

## `<ReferenceArrayInput>`

#### A spinner is displayed as long as:
 * we search for matching references and there are no references already linked to the ressource,
 * we search for matching references and also data for any references already linked to the ressource.
capture d ecran de 2018-02-12 16-22-43

#### An error is displayed if:
 * we have a server error return on the search of matching references and that there are no references already linked to the ressource,
 * we have a server error return on the search of matching references and no data for any of the references already linked to the ressource were found.
capture d ecran de 2018-02-12 16-24-19

#### A warning is displayed with a children's error when : 
 * we have a server error return on the search of matching references but we found the data for at least one reference already linked with the ressource,
 * We did not find the data for at least one of the references already associated with the ressource, but we have a valid return (even if empty) from the server on the search of matching references.
capture d ecran de 2018-02-12 16-25-11

File: docs/Inputs.md [hunk 21]
```diff
@@ -580,7 +580,8 @@ import { ReferenceInput, SelectInput } from 'admin-on-rest'
 </Admin>
 ```
 
-Set the `allowEmpty` prop when the empty value is allowed.
+Set the `allowEmpty` prop when you want to add an empty choice with a value of null in the choices list.
+Be careful, disabling `allowEmpty` does not mean that the input will be required. If you want to make the input required, you must add a validator as indicated in [Validation Documentation](./CreateEdit.html#validation). Enabling the `allowEmpty` props just adds an empty choice (with `null` value) on top of the options, and makes the value nullable.
```
Comment (djhi) [inline docs/Inputs.md]: s/Be careful, disabling/Disabling/

File: docs/Inputs.md [hunk 22]
```diff
@@ -695,7 +696,8 @@ import { ReferenceArrayInput, SelectArrayInput } from 'admin-on-rest'
...
```
Comment (djhi) [inline docs/Inputs.md]: s/Be careful, disabling/Disabling/

File: example/comments.js [hunk 23]
```diff
@@ -177,19 +178,20 @@ export const CommentEdit = props => (
 );
 
 const defaultValue = { created_at: new Date() };
+const required = value => (value ? undefined : 'Required');
```
Comment (djhi) [inline example/comments.js]: You can simply import the `required` function from `admin-on-rest`

File: docs/Tutorial.md [hunk 24]
```diff
@@ -249,11 +249,12 @@ const PostTitle = ({ record }) => {
     return <span>Post {record ? `"${record.title}"` : ''}</span>;
 };
 
+const required = value => (value ? undefined : 'Required');
```
Comment (djhi) [inline docs/Tutorial.md]: You can simply import the `required` function from `admin-on-rest`

File: src/mui/input/ReferenceArrayInput.spec.js [hunk 25]
```diff
@@ -2,202 +2,701 @@ import React from 'react';
...
+                    ...data,
+                    input: { value: [1, 2] },
+                    matchingReferences: { error: 'error' },
+                },
+                false,
+                'it is ready if linked reference data are not found, but the references fetch is finished with error.'
+            );
+        });
+
+        it('should claim an error if needed', () => {
```
Comment (djhi) [inline src/mui/input/ReferenceArrayInput.spec.js]: claim ?

File: src/mui/input/ReferenceArrayInput.spec.js [hunk 26]
```diff
@@ -2,202 +2,701 @@ import React from 'react';
...
+            assert.equal(ErrorElement.length, 0);
+            const MyComponentElement = wrapper.find('MyComponent');
+            assert.equal(MyComponentElement.length, 1);
+            assert.deepEqual(MyComponentElement.prop('meta'), {
+                error: '*fetch error*',
+                touched: true,
+            });
+        });
+
+        it('should send an error to the children if references were found and but selected references are not complete', () => {
```
Comment (djhi) [inline src/mui/input/ReferenceArrayInput.spec.js]: s/and but/and/

File: src/mui/input/ReferenceArrayInput.spec.js [hunk 27]
```diff
@@ -2,202 +2,701 @@ import React from 'react';
...
+            assert.equal(ErrorElement.length, 0);
+            const MyComponentElement = wrapper.find('MyComponent');
+            assert.equal(MyComponentElement.length, 1);
+            assert.deepEqual(MyComponentElement.prop('meta'), {
+                error: '*aor.input.references.many_missing*',
+                touched: true,
+            });
+        });
+
+        it('should send an error to the children if references were found and but selected references are empty', () => {
```
Comment (djhi) [inline src/mui/input/ReferenceArrayInput.spec.js]: s/and but/and/

File: src/mui/input/ReferenceArrayInput.spec.js [hunk 28]
```diff
@@ -2,202 +2,701 @@ import React from 'react';
...
+            );
+            const SpinnerElement = wrapper.find('ReferenceLoadingProgress');
+            assert.equal(SpinnerElement.length, 0);
+            const ErrorElement = wrapper.find('ReferenceError');
+            assert.equal(ErrorElement.length, 0);
+            const MyComponentElement = wrapper.find('MyComponent');
+            assert.equal(MyComponentElement.length, 1);
+        });
+
+        it('should not render enclosed component if allowEmpty is true', () => {
```
Comment (djhi) [inline src/mui/input/ReferenceArrayInput.spec.js]: should not render ?

File: src/mui/input/ReferenceArrayInput.spec.js [hunk 29]
```diff
@@ -2,202 +2,701 @@ import React from 'react';
...
+                    {...defaultProps}
+                    allowEmpty
+                    meta={{ touched: false }}
+                >
+                    <MyComponent />
+                </ReferenceArrayInput>
+            );
+
+            const myComponent = wrapper.find('MyComponent');
+            assert.notEqual(myComponent.prop('meta', undefined));
```
Comment (djhi) [inline src/mui/input/ReferenceArrayInput.spec.js]: `assert.deepEqual(myComponent.prop('meta'), { touched: false })` ?

File: src/mui/input/ReferenceInput.js [hunk 30]
```diff
@@ -2,15 +2,59 @@ import React, { Component } from 'react';
...
     crudGetMatching as crudGetMatchingAction,
 } from '../../actions/dataActions';
 import { getPossibleReferences } from '../../reducer/admin/references/possibleValues';
+import ReferenceLoadingProgress from './ReferenceLoadingProgress';
+import ReferenceError from './ReferenceError';
+import translate from '../../i18n/translate';
 
 const referenceSource = (resource, source) => `${resource}@${source}`;
 const noFilter = () => true;
+const getFinalLabel = (label, resource, source) =>
```
Comment (djhi) [inline src/mui/input/ReferenceInput.js]: useless function imo, why not just put the following where it is called ?
```jsx
label || `resources.${resource}.fields.${source}`
```

File: src/mui/input/ReferenceArrayInput.js [hunk 31]
```diff
@@ -2,14 +2,74 @@ import React, { Component } from 'react';
...
     crudGetMany as crudGetManyAction,
     crudGetMatching as crudGetMatchingAction,
 } from '../../actions/dataActions';
 import { getPossibleReferences } from '../../reducer/admin/references/possibleValues';
+import ReferenceLoadingProgress from './ReferenceLoadingProgress';
+import ReferenceError from './ReferenceError';
+import translate from '../../i18n/translate';
 
 const referenceSource = (resource, source) => `${resource}@${source}`;
+const getFinalLabel = (label, resource, source) =>
```
Comment (djhi) [inline src/mui/input/ReferenceArrayInput.js]: useless function imo, why not just put the following where it is called ?
```jsx
label || `resources.${resource}.fields.${source}`
```

File: src/mui/input/ReferenceArrayInput.spec.js [hunk 32]
```diff
@@ -2,202 +2,704 @@ import React from 'react';
...
-        const wrapper = shallow(
-            <ReferenceArrayInput {...defaultProps}>
-                <MyComponent />
-            </ReferenceArrayInput>
-        );
-        const MyComponentElement = wrapper.find('MyComponent');
-        assert.equal(MyComponentElement.length, 0);
-    });
+import {
+    getSelectedReferencesStatus,
```
Comment (fzaninotto) [inline src/mui/input/ReferenceArrayInput.spec.js]: It feels strange to test individual methods of a component. IMO, these are logic methods which should be placed out of the component. But we'll do that in another PR.

File: src/mui/input/ReferenceArrayInput.js [hunk 33]
```diff
@@ -172,33 +230,44 @@ export class ReferenceArrayInput extends Component {
...
-            label:
-                typeof label === 'undefined'
-                    ? `resources.${resource}.fields.${source}`
-                    : label,
+            label: translatedLabel,
             resource,
-            meta,
+            meta: {
+                ...meta,
+                error: dataStatus.warning,
```
Comment (fzaninotto) [inline src/mui/input/ReferenceArrayInput.js]: I'm afraid this may override actual errors from the filed level validator. Did you test it?

File: src/mui/input/ReferenceArrayInput.js [hunk 34]
```diff
@@ -232,44 +301,45 @@ ReferenceArrayInput.propTypes = {
...
     crudGetMatching: crudGetMatchingAction,
 })(ReferenceArrayInput);
 
-ConnectedReferenceInput.defaultProps = {
+ConnectedReferenceArrayInput.defaultProps = {
     addField: true,
 };
 
-export default ConnectedReferenceInput;
+export default translate(ConnectedReferenceArrayInput);
```
Comment (fzaninotto) [inline src/mui/input/ReferenceArrayInput.js]: Why do you decorate with translate here and not line 336?

File: src/mui/input/ReferenceArrayInput.js [hunk 35]
```diff
@@ -2,15 +2,72 @@ import React, { Component } from 'react';
...
 import { getPossibleReferences } from '../../reducer/admin/references/possibleValues';
+import ReferenceLoadingProgress from './ReferenceLoadingProgress';
+import ReferenceError from './ReferenceError';
+import translate from '../../i18n/translate';
 
 const referenceSource = (resource, source) => `${resource}@${source}`;
 
+export const getSelectedReferencesStatus = (input, referenceRecords) =>
+    !input.value || input.value.length === referenceRecords.length
+        ? 'ready'
```
Comment (fzaninotto) [inline src/mui/input/ReferenceArrayInput.js]: this should be a constant, same for 'incomplete' and 'empty'

File: src/mui/input/ReferenceArrayInput.js (hunk 34 above)
Comment (alexisjanvier) [reply src/mui/input/ReferenceArrayInput.js]: After review, it's working.

ezgif com-video-to-gif

ezgif com-video-to-gif 1

File: src/mui/input/ReferenceArrayInput.js (hunk 33 above)
Comment (alexisjanvier) [reply src/mui/input/ReferenceArrayInput.js]: After review, it's working.

ezgif com-video-to-gif

ezgif com-video-to-gif 1

**Comment by alexisjanvier** [comment]
@fzaninotto Review applied

File: src/mui/input/ReferenceArrayInput.js [hunk 36]
```diff
@@ -336,10 +344,10 @@ function mapStateToProps(state, props) {
 const ConnectedReferenceArrayInput = connect(mapStateToProps, {
     crudGetMany: crudGetManyAction,
     crudGetMatching: crudGetMatchingAction,
-})(ReferenceArrayInput);
+})(translate(ReferenceArrayInput));
```
Comment (djhi) [inline src/mui/input/ReferenceArrayInput.js]: would be better to use `recompose/compose` here, as done everywhere

**Comment by fzaninotto** [comment]
merged \o/

Now @djhi, you "just" have to port it to `next`...
